google/cloud/bigtable/test_proxy/cbt_test_proxy.cc

```diff
@@ -0,0 +1,392 @@
+  return grpc::Status();
+}
+
+// C++ client doesn't support closing which will make the client not accept new
+// calls for operation. As such, CloseClient() is implemented as no-op.
+grpc::Status CbtTestProxy::CloseClient(::grpc::ServerContext*,
+                                       testpb::CloseClientRequest const*,
+                                       testpb::CloseClientResponse*) {
+  return grpc::Status();
+}
+
+grpc::Status CbtTestProxy::RemoveClient(
+    ::grpc::ServerContext*, testpb::RemoveClientRequest const* request,
+    testpb::RemoveClientResponse*) {
+  std::lock_guard<std::mutex> lk(mu_);
+  auto client = connections_.find(request->client_id());
+  if (client == connections_.end()) {
+    return grpc::Status(
+        grpc::StatusCode::NOT_FOUND,
+        absl::StrCat("Client ", request->client_id(), " not found."));
+  }
+
+  connections_.erase(request->client_id());
+  return grpc::Status();
+}
+
+grpc::Status CbtTestProxy::ReadRow(::grpc::ServerContext*,
+                                   testpb::ReadRowRequest const* request,
+                                   testpb::RowResult* response) {
+  auto table = GetTableFromRequest(request->client_id(), request->table_name());
+  if (!table.ok()) return ToGrpcStatus(std::move(table).status());
+
+  auto const row =
+      table->ReadRow(request->row_key(), Filter(request->filter()));
+  if (row.ok()) {
+    if (row->first) {
+      Row row_val = row->second;
+      *response->mutable_row() = ConvertRowToV2(row_val);
+    } else {
+      GCP_LOG(ERROR) << "Received empty row: " << request->row_key();
+    }
+  }
+  *response->mutable_status() = ToRpcStatus(row.status());
+  return grpc::Status();
+}
+
+grpc::Status CbtTestProxy::ReadRows(::grpc::ServerContext*,
+                                    testpb::ReadRowsRequest const* request,
+                                    testpb::RowsResult* response) {
+  auto table = GetTableFromRequest(request->client_id(),
+                                   request->request().table_name());
+  if (!table.ok()) return ToGrpcStatus(std::move(table).status());
+
+  RowSet row_set;
+  for (RowKeyType const& row_key : request->request().rows().row_keys()) {
+    row_set.Append(std::string(row_key));
+  }
+  for (v2::RowRange const& row_range : request->request().rows().row_ranges()) {
+    row_set.Append(RowRange(row_range));
+  }
+  Filter const filter(request->request().filter());
+  auto const rows_limit = request->request().rows_limit();
+  RowReader reader =
+      rows_limit > 0 ? table->ReadRows(std::move(row_set), rows_limit, filter)
+                     : table->ReadRows(std::move(row_set), filter);
+
+  Status status;
+  for (auto const& row : reader) {
+    if (row.ok()) {
+      Row row_val = *std::move(row);
+      *response->add_row() = ConvertRowToV2(row_val);
+    } else {
+      status = row.status();
+      GCP_LOG(ERROR) << "Error reading row: " << row.status();
```

Comment:
Are these actually `ERROR`s? e.g. if we are given a bad row and expect the test proxy to not be able to read them, we probably just want to log `INFO`.

Reply:
Updated.